opentelemetry/src/common.rs

```diff
@@ -422,6 +424,55 @@ impl KeyValue {
     }
 }
 
+#[derive(Debug, Clone, Copy)]
+struct F64Hashable(f64);
+
+impl PartialEq for F64Hashable {
+    fn eq(&self, other: &Self) -> bool {
+        self.0.to_bits() == other.0.to_bits()
+    }
+}
+
+impl Eq for F64Hashable {}
+
+impl Hash for F64Hashable {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.0.to_bits().hash(state);
+    }
+}
+
+impl Hash for KeyValue {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.key.hash(state);
+        match &self.value {
+            Value::F64(f) => F64Hashable(*f).hash(state),
+            Value::Array(a) => match a {
+                Array::Bool(b) => b.hash(state),
+                Array::I64(i) => i.hash(state),
+                Array::F64(f) => f.iter().for_each(|f| F64Hashable(*f).hash(state)),
+                Array::String(s) => s.hash(state),
+            },
+            Value::Bool(b) => b.hash(state),
+            Value::I64(i) => i.hash(state),
+            Value::String(s) => s.hash(state),
+        };
+    }
+}
+
+impl PartialOrd for KeyValue {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Ord for KeyValue {
+    fn cmp(&self, other: &Self) -> Ordering {
+        self.key.cmp(&other.key)
+    }
+}
```

Comment:
I have some concerns about adding this behavior to the public API. The `KeyValue` struct is a generic data structure in our API, and its behavior should be consistent with how Rust handles (say) tuples. Specifically:

- Two `KeyValue` instances should be considered equal if both their key and value fields are equal.
- Both key and value should be considered for ordering comparisons.

The current proposal adds ordering functionality based solely on the `key` field. While this might suit our specific use case, such as using KeyValue in HashMap or HashSet where only keys are compared, it seems different from the intuitive behavior.
 
Maybe these are only my thoughts :)

Reply:
Can we add a (say) `KeyOnly` wrapper on top of `KeyValue` to enable using `KeyValue` in scenarios that need key-only ordering and comparison? I guess that can remain in the public API.

Reply:
thanks for raising this! It is a valid concern.
I think the doc for KeyValue can explain its behavior. It is OpenTelemetry Key,Value, not something else, so I think we can afford to do OpenTelemetry specific things.

`Ord` is currently only used for Metrics, but they could be used for other signals too, to achieve de-duplication.(deduping a sorted list can be done without hashset allocation). In those cases also, `Ord` would be based on Key only.

In other words, KeyValue is not required to match regular (key,value) tuples, and we need to document the intend in the doc. Otherwise, we could have just used (key,value) tuple itself, instead of implemented our own wrapper `KeyValue`.

Reply:
We also provide `KeyValue` attributes as argument in our Sampling API, and user may want to use some filtering logic in custom log-processors based on the attributes, and these sampling/filtering decision could require the default  `Ord` logic. 
But I don't have strong opinion here, if we can clearly document the behavior as part of this PR, we should be good.

Reply:
> could require the default Ord logic

That part, I am not sure! The fact that we are using a custom struct, and not (key,value) tuple, is to convey that it is *not* a normal tuple, and it's behavior is diff. from normal tuples. If a user intend to leverage a custom struct's trait implementations, I expect they read the doc for the same.

(I am yet to add docs comments in this PR, will add and update).

Reply:
I noticed this hashing and ordering is moved (back) to `metrics` directory, which means it would be available only if the metrics feature is enabled (though we enable it by default now) ? Should be fine, just confirming.

Reply:
true. I'll prepare a new PR to avoid dealing with conflict.

Reply:
>  moved (back) to metrics directory

But now in opentelemetry, not in sdk.